services/venice-router/src/main/java/com/linkedin/venice/router/api/path/VeniceMultiKeyPath.java

```diff
@@ -230,14 +240,18 @@ public int getLongTailRetryMaxRouteForMultiKeyReq() {
 
   @Override
   public boolean isLongTailRetryAllowedForNewRoute() {
+    // We are keeping both features for now. Retry is only allowed if both conditions pass:
+    // 1. retry attempts < longTailRetryMaxRouteForMultiKeyReq (if set).
+    // 2. we still have retry budget according to retry manager.
     if (longTailRetryMaxRouteForMultiKeyReq == -1) {
       // feature is disabled
-      return true;
-    }
-    if (longTailRetryMaxRouteForMultiKeyReq <= 0) {
+      return retryManager.isRetryAllowed();
+    } else if (longTailRetryMaxRouteForMultiKeyReq <= 0) {
       return false;
+    } else {
+      return (currentAllowedRetryRouteCnt.incrementAndGet() <= longTailRetryMaxRouteForMultiKeyReq)
+          && retryManager.isRetryAllowed();
```

Comment:
So the retry manager is enforcing in Route, which would result in some confusion:
1. Router is counting the original request.
2. Retry is being counted per route.
In theory, that is not fair or may not work as expected.
I think we need to align these two measurements, and ideally #1 should count routes of the original request instead.
The total number of requests is critical when evaluating retry storm.